src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs

```diff
@@ -1004,14 +1122,731 @@ static void Main()
+                //         F1([1], [2]);
+                Diagnostic(ErrorCode.ERR_BadArgType, "[1]").WithArguments("1", "collection literals", "dynamic").WithLocation(10, 12),
+                // (11,12): error CS1503: Argument 1: cannot convert from 'collection literals' to 'D'
+                //         F2([3], [4]);
+                Diagnostic(ErrorCode.ERR_BadArgType, "[3]").WithArguments("1", "collection literals", "D").WithLocation(11, 12),
+                // (12,12): error CS1503: Argument 1: cannot convert from 'collection literals' to 'E'
+                //         F3([5], [6]);
+                Diagnostic(ErrorCode.ERR_BadArgType, "[5]").WithArguments("1", "collection literals", "E").WithLocation(12, 12));
+        }
+
+        [Fact]
+        public void TypeInference_30()
+        {
+            string source = """
+                delegate void D();
+                enum E { }
+                class Program
+                {
+                    static void F1<T>(dynamic[] x, T[] y) { }
+                    static void F2<T>(D[] x, T[] y) { }
+                    static void F3<T>(E[] x, T[] y) { }
+                    static void Main()
+                    {
+                        F1([1], [2]);
+                        F2([null], [4]);
+                        F3([default], [6]);
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void TypeInference_31()
+        {
+            string source = """
+                class Program
+                {
+                    static void F<T>(T[] x) { }
+                    static void Main()
+                    {
+                        F([null]);
+                        F([Unknown]);
+                        F([Main()]);
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (6,9): error CS0411: The type arguments for method 'Program.F<T>(T[])' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         F([null]);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "F").WithArguments("Program.F<T>(T[])").WithLocation(6, 9),
+                // (7,12): error CS0103: The name 'Unknown' does not exist in the current context
+                //         F([Unknown]);
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "Unknown").WithArguments("Unknown").WithLocation(7, 12),
+                // (8,9): error CS0411: The type arguments for method 'Program.F<T>(T[])' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         F([Main()]);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "F").WithArguments("Program.F<T>(T[])").WithLocation(8, 9));
```

Comment:
Feels like an additional error for `Main()` outside an expression statement would be nice, since it's always going to be broken in this context, but I wouldn't spend much time on it.